File: EPIJudge/epi_judge_cpp/lowest_common_ancestor_with_parent.cc
#include "binary_tree_with_parent_prototype.h"
#include "test_framework/binary_tree_utils.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"

int height(BinaryTreeNode<int> *node) {
    int h = 0;
    while (node != nullptr) {
        h++;
        node = node->parent;
    }
    return h;
}

BinaryTreeNode<int>* Lca(const unique_ptr<BinaryTreeNode<int>>& node0,
                         const unique_ptr<BinaryTreeNode<int>>& node1) {
    auto *it0 = node0.get();
    auto *it1 = node1.get();
    int h0 = height(it0);
    int h1 = height(it1);
    if (h1 > h0) std::swap(it0, it1);
    int diff = abs(h1 - h0);
    while (diff--) it0 = it0->parent;
    while (it0 != it1) {
        it0 = it0->parent;
        it1 = it1->parent;
    }
    return it0;
}

int LcaWrapper(TimedExecutor& executor,
               const unique_ptr<BinaryTreeNode<int>>& tree, int key0,
               int key1) {
  const unique_ptr<BinaryTreeNode<int>>& node0 = MustFindNode(tree, key0);
  const unique_ptr<BinaryTreeNode<int>>& node1 = MustFindNode(tree, key1);

  auto result = executor.Run([&] { return Lca(node0, node1); });

  if (!result) {
    throw TestFailure("Result can not be nullptr");
  }
  return result->data;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "tree", "key0", "key1"};
  return GenericTestMain(args, "lowest_common_ancestor_with_parent.cc",
                         "lowest_common_ancestor.tsv", &LcaWrapper,
                         DefaultComparator{}, param_names);
}

File: EPIJudge/epi_judge_python/lowest_common_ancestor_with_parent.py
import functools
from typing import Optional

from binary_tree_with_parent_prototype import BinaryTreeNode
from test_framework import generic_test
from test_framework.binary_tree_utils import must_find_node
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook


def lca(node0: BinaryTreeNode,
        node1: BinaryTreeNode) -> Optional[BinaryTreeNode]:
    def depth(node):
        total = 0
        while node:
            total += 1
            node = node.parent
        return total

    d0 = depth(node0)
    d1 = depth(node1)

    if d0 < d1:
        d0, node0, d1, node1 = d1, node1, d0, node0

    while d1 != d0:
        d0 -= 1
        node0 = node0.parent

    while node0 != node1:
        node0 = node0.parent
        node1 = node1.parent

    return node0


@enable_executor_hook
def lca_wrapper(executor, tree, node0, node1):
    result = executor.run(
        functools.partial(lca, must_find_node(tree, node0),
                          must_find_node(tree, node1)))

    if result is None:
        raise TestFailure('Result can\'t be None')
    return result.data


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('lowest_common_ancestor_with_parent.py',
                                       'lowest_common_ancestor.tsv',
                                       lca_wrapper))
